### ISSUE TITLE
Improve lore2sphinx buildbot results for `projects/mail`

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4572]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)| @khorn reported|
|-|-|
|Trac ID|trac#4572|
|Type|enhancement|
|Created|2010-07-11 07:02:23Z|
|Branch|https://github.com/twisted/twisted/tree/lore2sphinx-mail-4572|

like [#4566](https://github.com/twisted/twisted/issues/4566), except for `projects/mail`

Attachments:

* [index.rst](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/22b/22b1e9639ca39a26e8585c13f7f4237b6ab8a3ea/51074e66e615ec7bc5afa3b4022a883cd94c0dec.rst) (436 bytes) - added by jdb on 2010-07-23 09:25:24Z - 
* [index.2.rst](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/22b/22b1e9639ca39a26e8585c13f7f4237b6ab8a3ea/82cee84fe004b64573f322e4974b026bd1036ea2.rst) (526 bytes) - added by jdb on 2010-07-23 09:26:36Z - mail/examples/index.rst
* [smtpclient.rst](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/22b/22b1e9639ca39a26e8585c13f7f4237b6ab8a3ea/e66bd0c90d0e60710851cbf5534548447223c400.rst) (29549 bytes) - added by jdb on 2010-07-23 09:27:43Z - Corrected verison of the SMTP tutorial
* [smtpclient-old.rst](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/22b/22b1e9639ca39a26e8585c13f7f4237b6ab8a3ea/5c52313a20d26b3db96c0aee51e29967eb49b3af.rst) (29954 bytes) - added by jdb on 2010-07-23 09:28:05Z - Open version of the SMTP tutorial
* [smtpclient.xhtml.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/22b/22b1e9639ca39a26e8585c13f7f4237b6ab8a3ea/7a530da692e77f9b44683c96703c6efec1e3b43a.patch) (17365 bytes) - added by jdb on 2010-07-23 13:49:35Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4572 4572
type__enhancement enhancement
reporter__khorn khorn
priority__normal normal
milestone__Lore_to_Sphinx Lore to Sphinx
branch__branches_lore2sphinx_mail_4572 branches/lore2sphinx-mail-4572
branch_author__jdb jdb
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1278831743000000 1278831743000000
changetime__1283702738000000 1283702738000000
version__None None
owner__ 
cc__jdb
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jdb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|jdb commented|
|-|-|



The mail documentation chunk is composed of three rst files: two
short listings and one sizeable tutorial::

  doc/projects/mail
  |-- examples
  |   `-- index.rst
  |-- index.rst
  `-- tutorial
      `-- smtpclient
          `-- smtpclient.rst

The transformation is almost correct, there are two bugs that show in HTML (they show especially in the file mail/tutorial/smtpclient/smtpclient.rst):

#. there is a space missing a space before an italic or preformatted
   markup (not always reproduceable). This bug is a problem as it 
   disables the markup.

   ex: look for the text : "use by``twistd`` ." The original xhtml was
   "use by <code>twistd</code>.</p>". Because of the missing space, the
   markup is incorrect and the four backquotes can be read untransformed
   in html.

#. there is always extra space in the middle of:

   - an underline or preformatted string, or a link: ("\:ref:\`link` ,")

   - a dot, comma, parenthesis, colon, apostrophy

   ex: look for the text : "\*application service\* ."

Often both bugs happen on the same chunk of text. Here are two other minor bugs of the rst file which do not show in HTML: 

- four or even six unneeded blank lines between text blocks while 
  one was enough 

- weird paragraph *fill* (lines sometimes 30 characters long, sometimes 150)

Attached are the three rst files corrected by hand. Also the version buggy version of the tutorial is attached. 

Question: is there a repository of restructured text that I can commit the version corrected by hand?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Apparently [#4566](https://github.com/twisted/twisted/issues/4566) didn't explain that the idea here is not to hand-fix the rst output, but to fix the Lore _input_ so that the output comes out looking better.

**Comment by twisted-trac** [comment]
|<img alt="jdb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|jdb commented|
|-|-|

Replying to [exarkun](#note_3):
> Apparently [#4566](https://github.com/twisted/twisted/issues/4566) didn't explain that the idea here is not to hand-fix the rst output, but to fix the Lore _input_ so that the output comes out looking better.


Attached is a patch to the smtpclient xhtml/lore input where no text markup follows a newline character. This seems to fix the problem of rst markup being visible in HTML because of a missing space in the rst file. 

Concerning the problem of the extra space created between a markup and a final dot, I haven't found a way yet to "adapt" the xhtml. 

Does someone have trick for getting rid of the extra space?

It is also possible to rephrase each sentence so that no markuped expressions is in the end of a sentence (we do not want to do this I guess). It is also possible to leave this minor bug unfixed and fix it later on in the rst source.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra removed owner|
|-|-|

Putting the patch up for review.

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|Screwtape commented|
|-|-|

Replying to [jdb](#note_4):
> Concerning the problem of the extra space created between a markup and a final dot, I haven't found a way yet to "adapt" the xhtml. 

Yeah, I believe the plan is to fix the generated ReST markup after we throw the switch and make the Sphinx docs the official ones.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_7"></a>|@khorn commented|
|-|-|

Yes, that is indeed the plan.  Once this set of tickets has been completed, we'll convert to Sphinx and then clean up the remaining issues with another set of tickets.  There are some issues which are just very difficult to translate automatically (as jdb has discovered).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [29968]) Branching to 'lore2sphinx-mail-4572'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [29969]) Apply smtpclient.xhtml.patch

refs [#4572](https://github.com/twisted/twisted/issues/4572)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [29970]) Correct two more whitespace issues which cause lore2sphinx to emit bogus restructured text

refs [#4572](https://github.com/twisted/twisted/issues/4572)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  This looks good, except for a couple extra changes I made above.  Merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [29972]) Merge lore2sphinx-mail-4572

Author: jdb
Reviewer: exarkun
Fixes: [#4572](https://github.com/twisted/twisted/issues/4572)

Various changes to the Twisted Mail documentation to improve the results
of the lore2sphinx tool.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
